src/main/drivers/vtx_common.c

```diff
@@ -25,14 +25,43 @@
 #include "platform.h"
 #include "build/debug.h"
 
-#if defined(VTX_COMMON)
+#include "drivers/vtx_common.h"
+
+#include "fc/config.h"
 
-#include "vtx_common.h"
+#include "io/vtx_common.h"
```

Comment:
We need to maintain proper separation between the IO layer and the drivers layer. In particular drivers cannot call upwards - this is a hard and fast rule.

This means that drivers are not allowed to include anything from the IO layer. So this means the declaration:
```
typedef struct vtxSettingsConfig_s {
    uint8_t band;       // 1=A, 2=B, 3=E, 4=F(Airwaves/Fatshark), 5=Raceband
    uint8_t channel;    // 1-8
    uint8_t power;      // 0 = lowest
    uint16_t freq;      // sets freq in MHz if band=0
} vtxSettingsConfig_t;
```
should be moved to `drivers/vtx_common.h`, whereas the declaration `PG_DECLARE(vtxSettingsConfig_t, vtxSettingsConfig);` stays in `io/vtx_common.h`. Also any constants from `io/vtx_common.h` that are required by the driver should be moved to `drivers/vtx_common.h`.

Sorry if I did not explain this properly before.

Note that "drivers must not call upward" is a very important rule to enable maintenance and understandability of code, as well as helping to avoid bug.s

Reply:
To me, it feels like the structure should be declared alongside the PG, since they are so tightly coupled.  The drivers should just be a consumer

Reply:
In a sense, the io would be "handing down" the PG to the drivers

Reply:
Does it make sense to roll all of this driver abstraction to the io layer?

Reply:
Once `vtxCommonInit` and `vtxCommonProcess` are moved to the IO layer the driver won't need to reference `vtxSettingsConfig`, so the question becomes moot.

And yes, I think it is correct to move these things to the IO layer. It enforces proper separation of responsibilities in the code. This makes the code easier to understand and maintain. In general drivers should be very simple things with limited responsibility.

Reply:
No, there should be a io-device independant API in the driver layer. Io-device specifics in the io-layer.

Reply:
@AndersHoglund , I agree with your comment, but not sure why you prefixed it with "No,".

Reply:
"No" as to answer the Q "Does it make sense to roll all of this driver abstraction to the io layer?"
But maybe I misunderstood the Q and what "...this driver.." referred to.

Reply:
Well, I think it's tautological that he driver abstraction belongs in the driver layer. However `vtxCommonProcess` is certainly not part of the driver abstraction - it's the VTX task that is run by the scheduler. Also `vtxCommonInit` is not part of the driver abstraction.

Reply:
Would it be inaccurate to assume that the behavior of an API is to facilitate IO?  Not debating the suggestions, just throwing it out there.